### PR TITLE
feat(autocomplete): cache elements for a specific lifetime in milliseconds

### LINKED ISSUE #2705: md-autocomplete: Add a cache lifetime attribute
`md-autocomplete` has an `md-no-cache` option to completely disable the cache. But it'd be nice to meet in the middle and let users specify a cache lifetime for results.

A use case would be searching through items which a team of people can create/edit. Caching for 30s or even a few minutes is a lot better then no cache because while searching a cache is helpful. But things aren't likely to change within 5 seconds depending on your team size. Also if we cache for ever then you would have to refresh the page to see new items from the search.

### PR DESCRIPTION
**Pull request by DevVersion**
Fixes #2705

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
@robertmesserle - not sure I like this extension to mdAutoComplete. Can you review?

**Comment by robertmesserle** [comment]
@ThomasBurleson It feels clunky and a bit specific, but the use-case brought up in #2705 is a valid one.  I'm fine with this feature being added.

File: src/components/autocomplete/js/autocompleteController.js [hunk 1]
```diff
@@ -684,6 +686,20 @@ function MdAutocompleteCtrl ($scope, $element, $mdUtil, $mdConstant, $mdTheming,
...
   /**
+   * Removes the given term after the specified cache lifetime
+   * @param term
+   */
+  function removeAfterLifetime(term) {
+    if ($scope.cacheLifetime && lifetimeQueue.indexOf(term) === -1) {
+      lifetimeQueue.push(term);
+      $timeout(function() {
+        cache[term] = null;
+        lifetimeQueue.splice(lifetimeQueue.indexOf(term), 1);
```
Comment (robertmesserle) [inline src/components/autocomplete/js/autocompleteController.js]: Should we store the index since it is referenced more than once?

Comment (robertmesserle) [inline src/components/autocomplete/js/autocompleteController.js]: Or would a hashmap make more sense for this with boolean true/false values?

Comment (DevVersion) [inline src/components/autocomplete/js/autocompleteController.js]: The thing is, the index will change if a previous queue task will remove an item from the list.

So we need to retrieve the index on removing.
I think indexOf is a good solution.

**Comment by DevVersion** [comment]
@robertmesserle What do you mean with clunky? Is there anything I can improve? :)

File: src/components/autocomplete/js/autocompleteController.js (hunk 1 above)
Comment (robertmesserle) [inline src/components/autocomplete/js/autocompleteController.js]: Ah, right - didn't notice it was inside a timeout.  Disregard. =p

**Comment by robertmesserle** [comment]
@DevVersion Oh, no, the implementation is fine.  Clunky probably wasn't the right word, but in the past, I was on the fence about this feature because it felt very edge-casey.

**Comment by DevVersion** [comment]
@ThomasBurleson Rebased to master and resolved appeared merge conflict (from [here](https://github.com/angular/material/blob/52a519ead04f2faeba227fbc51a7214253cbf2ba/src/components/autocomplete/js/autocompleteController.js#L22).

**Comment by ThomasBurleson** [comment]
Solution is too specifc and not central to ngMaterial. Rejecting and closing.

**Pull request closed without merging**